### DISCUSSION TITLE
Is there a need to add an underline button to rich text?

### DISCUSSION THREAD

**Discussion by t-hamano** [WordPress/gutenberg#43616] [General]
I noticed that of all the format types in the core, only the underline is not showing up in the toolbar or dropdown.
Below are the results of the survey:

| Format Type        | Shown in the toolbar? | Mentioned in shortcut modal? | Have a shortcut? (On Windows) |
| ------------------ | --------------------- | ---------------------------- | ----------------------------- |
| core/bold          | ✅ (toolbar)           | ✅                            | ✅ ( `Ctrl + B` )              |
| core/code          | ✅ (toolbar)           | ✅                            | ✅ ( `Shift + Alt + X` )       |
| core/image         | ✅ (dropdown)          | -                            | -                             |
| core/italic        | ✅ (toolbar)           | ✅                            | ✅ ( `Ctrl + I` )              |
| core/link          | ✅ (toolbar)           | ✅                            | ✅ (`[[`)                      |
| core/striktthrough | ✅                     | ✅                            | ✅ ( `Shift + Alt + D` )       |
| core/underline     | ⚠️                     | ✅                            | ✅                             |
| core/text-color    | ✅                     | -                            | -                             |
| core/subscript     | ✅                     | -                            | -                             |
| core/superscript   | ✅                     | -                            | -                             |
| core/keyboard      | ✅                     | -                            | -                             |

Looking at this table alone, it seems odd that only the underline doesn't have a button, but this feature is not a deprecated "Underline" element, but simply wraps in a `span` tag with an underline style.

#13117 is the PR with an additional underline shortcut by `Cmd + U`.
I also consulted the relevant issue, and it is my understanding that the underline style does not explicitly provide a button because it could be confused with the inline style.

Personally, I sometimes want to give an underline that is not a link, but is there any possibility of a button being added in the future?

**Comment by jasmussen** [comment]
I do not have the full history, but it is my understanding that the underline button is intentionally not surfaced because underlined text carries with it a risk of being confused with a hyperlink. This is especially important for low-vision or color-blind users, for whom usually colored hyperlinks may not be sufficiently differentiated from underlined text. Having the shortcut available, but the button not surfaced, seems like an acceptable balance to me.

**Reply by t-hamano** [reply]
Thank you for the detailed explanation.
That certainly makes sense.

**Reply by scruffian** [reply]
The problem is that in some themes, underlines are explicitly removed from links for example: https://github.com/Automattic/themes/blob/ca887cd3f72c1b755c0b5b6d2fe2c85d7597411c/pendant/theme.json#L378

In this case if I want an underlined link outside the Post Content the only way to do it is using a keyboard shortcut.

**Comment by ramonjd** [comment]
Coming here from a WordPress.org forum question. https://wordpress.org/support/topic/why-is-there-no-option-to-underline-links-in-the-block-editor/

Is this still relevant?

How strong is the accessibility argument here? 

What if the underline button were in the block tool drop down and on not in the toolbar itself (next to **Bold** and _Italics_)?

If it's a definite "no" I think we could close this given https://github.com/WordPress/gutenberg/issues/47484 was closed as well

**Reply by jasmussen** [reply]
I would still not show it. It's not great practice to underline hypertext, because it confuses with people of low vision who rely on the shape to identify links. The fact that you _can_ do it, via ⌘U, I'd suggest is plenty. But it's also not a strong opinion, so I'm happy to defer to consensus otherwise. Though I would say if we do show it, show it in the inline formatting dropdown, not in the main formatting bar. 

There has been separate discussions around converting that inline formatting dropdown to Menu V2 which supports flyouts. One of those flyouts could be special characters. Another could be additional inline formatting.

**Reply by t-hamano** [reply]
Having reviewed the past discussions again, it still consistently appears not to be recommended from an accessibility standpoint.

- https://make.wordpress.org/core/2016/10/28/editor-changes-in-4-7/#:~:text=Underlining%20is%20a%20bad%20practice%20as%20readers%20can%20confuse%20it%20with%20links
- https://github.com/WordPress/gutenberg/issues/12814#issuecomment-446609999
- https://github.com/WordPress/gutenberg/pull/1585#issuecomment-311942564

**Reply by ramonjd** [reply]
> from an accessibility standpoint

Thank you for finding those links @t-hamano. 

I followed up on them, and find the arguments unconvincing in the face of explicit user demand. "Why do folks keep asking for this?" is what we should be asking

The [4.7 post](https://make.wordpress.org/core/2016/10/28/editor-changes-in-4-7/) is cited as the authority, but it has no sources - it reads like a decision and not evidence of why we don't offer something that users want (and something that they're used to from Wix, Word, Google Docs and Slack etc etc)

Sorry to bang on about this, but I want us to do things for good reasons, and it seems to me we've been treating that post from 2016 and other speculative comments as a "standard". 

Hiding the button doesn't stop underlining. Cmd+U still works. So does the native `formatUnderline` input event. So does pasting from Word. Folks know what a link is and what an underline is, and I think we can trust authors to make decisions for themselves.

I think the drop down is, at least for now, a practical place for this tool. 

I think I need a tea and a lie down 🤣 

> if we do show it, show it in the inline formatting dropdown, not in the main formatting bar.

I'll get a PR up for it. I'm 100% fine if it gets shot down again, but I'd be raising the same points above.

**Reply by ramonjd** [reply]
P.S. thanks for entertaining my objections, and for closing this discussion.

> I'll get a PR up for it. I'm 100% fine if it gets shot down again, but I'd be raising the same points above.

I'll sleep on it, and if I'm still worked up I'll do this 😄

**Reply by t-hamano** [reply]
> > I'll get a PR up for it. I'm 100% fine if it gets shot down again, but I'd be raising the same points above.
> 
> I'll sleep on it, and if I'm still worked up I'll do this 😄 

Regarding this, I noticed that a PR has already been submitted. https://github.com/WordPress/gutenberg/pull/78561

We might be able to discuss it with more contributors there.

**Reply by jasmussen** [reply]
As noted, while I have an opinion, it's not a strong or blocking one, and I did propose submenu flyouts if we want to expand the range of controls. But I wanted to respond to this:

> I followed up on them, and find the arguments unconvincing in the face of explicit user demand. "Why do folks keep asking for this?" is what we should be asking

It's always a careful balance when listening to user feedback to try and find out what the user is trying to accomplish, vs. what they are asking for, as the two are not always the same. In that vein, I would love to understand good use cases for _inline_ underline styles. Note, _global styles-based_ underlines I think of as different: for example when styling links as part of [an elements section](https://github.com/WordPress/gutenberg/pull/81645), naturally we will need to provide underline as a style option. The default one, even. 

Crucially, just like how we did when allowing justified text on paragraphs, when it's a global style we can provide a notice explaining that while it is there for you to choose, it will affect legibility:

<img width="353" height="182" alt="image" src="https://github.com/user-attachments/assets/3455cbe0-493f-4b16-afa3-37d4f9eb11fa" />

And for justified-text, we also kept hearing from users: I need justified text. We also did the research: _why_ do users need justified text? And the answer is: in many newsrooms or for certain school assignments, there's a manual of style that mandates justified text. This helped inform the solution: if it's a manual of style, you set it _once_, globally, and then all your paragraphs are justified. But what users were asking for, was adding justified text in the block toolbar dropdown:

<img width="393" height="245" alt="image" src="https://github.com/user-attachments/assets/7decfdac-049f-4898-88f0-a37d0474ec97" />

That would have been a mistake, because then users would have had to edit every post on their site, manually selecting and aligning every single paragraph, and applying a per-block, per-post style, for something that was meant to be a single style set in one place. This is why it's so important to understand that there's always nuance.

With underlined text, for example, I wonder if there's a similar situation: people ask for it as an inline style when in fact they need it for underlined headings, which would almost certainly be better set, alongside a yellow notice, in global styles.

**Reply by ramonjd** [reply]
Thank you for indulging my rant and taking the time to reply. 

Always happy to be put in my place 😄 

Maybe there's a middle ground as @jasmussen describes.

It could come down to a theme choice. From the [forum](https://make.wordpress.org/core/2016/10/28/editor-changes-in-4-7/) report:

> Some themes underline links automatically, but many do not.

**Reply by jasmussen** [reply]
Didn't mean to put anyone in their place, to be clear.

> Some themes underline links automatically, but many do not.

Ah, see _that's_ an important piece. Link styles. We absolutely need something like the justification control, but for link styles, in the Element section, so that if a theme _doesn't_ underline links, users can correct that. And we can show that warning notice for themes that do this.
